src/storage/src/source/timestamper.rs

```diff
@@ -0,0 +1,325 @@
+            if !PartialOrder::less_equal(&self.write_upper, &self.read_progress) {
+                // Use intentionally wrong value of `expected_upper` in order to update the upper for next iteration.
+                let empty: [(((), PartitionId), Timestamp, Diff); 0] = [];
+                self.write_upper = self
+                    .write_handle
+                    .compare_and_append(
+                        empty,
+                        Antichain::from_elem(Timestamp::minimum()),
+                        Antichain::from_elem(Timestamp::minimum() + 1),
+                    )
+                    .await
+                    .expect("Timestamper upper update CAA")
+                    .with_context(|| "Timestamper upper update compare and append")?
+                    .map_err(|Upper(actual_upper)| actual_upper)
+                    .expect_err("Antichain was advanced past min on initialization");
+
+                continue;
+            }
+
+            // Compact bindings we've read in and see if we're able to assert a timestamp for any of the input offsets
+            for (partition, offset) in observed_max_offsets.iter() {
+                if matched_offsets.contains_key(partition) {
+                    continue;
+                }
+                let bindings = match self.persisted_timestamp_bindings.get_mut(&partition) {
+                    Some(bindings) => bindings,
+                    None => continue,
+                };
+                if bindings.is_empty() {
+                    continue;
+                }
+
+                // Compact as able, relying on all messages being in ascending offset order
+                while bindings.len() > 1
+                    && bindings
+                        .front()
+                        .expect("always at least one binding per partition")
+                        .1
+                        < *offset
+                {
+                    let (_old_timestamp, old_max_offset) = bindings.pop_front().unwrap();
+                    let (_timestamp, incremental_offset) = bindings.front_mut().unwrap();
+                    *incremental_offset += old_max_offset;
+                }
+
+                let (timestamp, max_offset) = bindings
+                    .front()
+                    .expect("always at least one binding per partition");
+                if *offset <= *max_offset {
+                    let old_value =
+                        matched_offsets.insert(partition.clone(), (*timestamp, *offset));
+                    assert_eq!(old_value, None);
+                } else {
+                    assert_eq!(bindings.len(), 1);
+                    // Unable to match any more from this partition
+                    break;
+                }
+            }
+
+            // TODO(#12267): Properly downgrade `since` of the timestamp collection based on the source data collection.
+
+            // XXX: should clamp to round timestamp_frequency??
```

Comment:
Yeah that would be a good idea to reduce number of unique timestamps in the system. We also used to have some code that defended againsts the clock going backwards. `NowFn` doesn't make any guarantee to the behaviour of its return values

Reply:
it's not monotonic?? okay